Improve stack shim performance

kenkangxgwe opened the pull request:
Use nested tuple instead of AppendTo to improve the top-level expression abstraction.
The old implementation takes O(n) for Push, and Pop, and O(1) for Normal,
while the new implementation takes O(1) for Push, and Pop, and O(n) for Normal.
This accelerates the `abstractTopLevel` function since it contains n pushes and pops and m normalizations of size n/m.
Thus, this change improves it from O(n^2) to O(n), which is efficient enough to remove the $TopLevelExpressionLimit.

This can be verified by running following tests:

AppendTo
![image](https://user-images.githubusercontent.com/8650023/79034847-f5942580-7b6d-11ea-8a23-d79ca860b4f9.png)

Nested Tuple:
![image](https://user-images.githubusercontent.com/8650023/79035118-342adf80-7b70-11ea-95d2-1b00e00ccf58.png)

__Update:__ thanks to @szhorvat's reminding me of using a custom head, the `Normal` is now ~30x faster.

![image](https://user-images.githubusercontent.com/8650023/79193010-a30d7000-7dde-11ea-9108-1ea97f6f9d61.png)

kenkangxgwe commented:
Unfortunately, I don't know how to run the tests in the `Tests` folder, so I didn't verify that this commit truly accelerates `abstractTopLevel`. Can you give details on how can I run those tests?

bostick commented:
I think this will be a very good change! I did not spend time optimizing the Stack shim, so thank you for looking at this!

For running tests, you can use Workbench or you can use the MUnit package directly:

```

<<MUnit`

TestRun["/Users/brenton/development/stash/COD/codeparser/Tests/TestSuite.mt"]

Starting test run "Test Report: TestSuite.mt"
...........................................................................................................................................................................................................................................Starting test section DotNegative

Ending test section
.................................................Starting test section ImplicitPlus

Ending test section
..............Starting test section TwoWayRule
....
Ending test section
........................................................................................................................................................................................................................................................................................
Tests run: 584
Failures: 0
Messages Failures: 0
Skipped Tests: 2
Errors: 0
Fatal: False
```

You will see a lot of messages printed out because I have not silenced them. There is a lot of testing invalid syntax and that generates messages.

But I think existing tests are not so good for testing this timing improvement.

Maybe some kind of very small benchmark could be written for timing tests.

bostick commented on CodeParser/Kernel/Shims.wl:
```diff
@@ -28,37 +28,44 @@ setupShims[] := (
-          tmp = stackVal[[-1]];
-          stackVal = Drop[stackVal, -1];
-          tmp
+        Block[{tmp},
```
```suggestion
        Module[{tmp},
```

I know that Block is faster, but I want to maintain a consistent style.

bostick commented on CodeParser/Kernel/Shims.wl:
```diff
@@ -28,37 +28,44 @@ setupShims[] := (
-      stack /: Normal[stack] := stackVal;
+      stack /: Normal[stack] := (
+        Nest[(Sow[#[[1]]]; #[[2]])&, stackVal, stackDepth]
+        // Reap // Last // Last // Reverse
```
I would prefer doing  `Reap[...][[2, 1]]`

I don't like the style of cascading `//`

szhorvat commented:
@bostick It's not nice that a new `System` symbol is created M < 12.1 (`CreateDataStructure`). If multiple packages do such things, there will be conflicts. Maybe, as part of this change, the whole thing could be better hidden inside of the CodeParser package?

szhorvat commented:
@kenkangxgwe Just an idea: I usually implement this kind of data structure with a custom head instead of with `List`. So instead of having something like `{oldStack, value}`, I use `tuple[oldStack, value]` where `tuple` is a package-private symbol. Then it is safe to simply `Flatten[stack, Infinity, tuple]` to get it in list form (the `Normal` operation).

bostick commented:
@szhorvat Yes, messing with System` is not so nice. Can perhaps be addressed in another PR.

I vote for custom head of `cons` or something like that, to evoke the cons cell nature of the data structure. But keeping `List` or changing it to anything is fine.

kenkangxgwe commented:
@szhorvat , thanks for the great idea. I was thinking that I missed something when I implemented the `Normal`, and you reminded me. It is the custom header that Flatten can deal with it elegantly instead of using `Nest` to peel off the list layer by layer. It is still O(n) but much faster. I've updated the benchmark in OP.

------

@bostick , I have ran the test calling the `MUnit`` package. I found that there are 8 failures on both 12.1 and 11.3. Do all tests pass on your side? I consider these are not caused by the stack shim.

![image](https://user-images.githubusercontent.com/8650023/79196208-78beb100-7de4-11ea-8067-4ac887de20d1.png)

bostick commented on CodeParser/Kernel/Shims.wl:
```diff
@@ -28,37 +28,43 @@ setupShims[] := (
 
       stack /: stack["Push", expr_] := (
-        AppendTo[stackVal, expr];
+        stackVal = stackCons[expr, stackVal];
```
If you do: `stackVal = stackCons[stackVal, expr];` ...

bostick commented on CodeParser/Kernel/Shims.wl:
```diff
@@ -28,37 +28,43 @@ setupShims[] := (
-          tmp
+          If[stackDepth != 0,
+            stackDepth -= 1;
+            {tmp, stackVal} = List @@ stackVal;
```
... and also do `{stackVal, tmp} = List @@ stackVal` ...

bostick commented on CodeParser/Kernel/Shims.wl:
```diff
@@ -28,37 +28,43 @@ setupShims[] := (
 
-      stack /: Normal[stack] := stackVal;
+      stack /: Normal[stack] := (
+        Reverse[Flatten[{stackVal}, Infinity, stackCons]]
```
... then the `Reverse` here can be removed

bostick commented on CodeParser/Kernel/Shims.wl:
```diff
@@ -28,37 +28,43 @@ setupShims[] := (
         ];
 
-      stack /: stack["Peek"] := stackVal[[-1]];
+      stack /: stack["Peek"] := First[stackVal];
```
... and do `      stack /: stack["Peek"] := Last[stackVal];` ...

bostick commented:
@kenkangxgwe I opened a separate issue for getting tests running: https://github.com/WolframResearch/codeparser/issues/6

bostick commented:
@kenkangxgwe Since we are obviating the need for `$TopLevelExpressionLimit`, can you just remove occurrences of it from `Shims.wl` and `Abstract.wl` ?

bostick commented:
Thanks for making the first public PR for codeparser @kenkangxgwe !
